docs/puzzles/day16.md

```diff
@@ -1,11 +1,572 @@
+```
+
+`Packet.Literal` will represent simple literal packets, that contain only a
+value. We are using Long just in case based on the experience with previous
+Advent of Code puzzles.
+
+`Packet.Operator` will represent all the other operators that can contain other
+packets.
+
+Now we need to map our input to these structures that we defined.
+
+Let's start by mapping out hexadecimal input to list of chars that we can
+analyse easier when checking for packets:
+
+```scala
+val hexadecimalMapping =
+  Map(
+    '0' -> "0000",
+    '1' -> "0001",
+    '2' -> "0010",
+    '3' -> "0011",
+    '4' -> "0100",
+    '5' -> "0101",
+    '6' -> "0110",
+    '7' -> "0111",
+    '8' -> "1000",
+    '9' -> "1001",
+    'A' -> "1010",
+    'B' -> "1011",
+    'C' -> "1100",
+    'D' -> "1101",
+    'E' -> "1110",
+    'F' -> "1111"
+  )
+
+def parse(input: String) =
+  val byteInput = input.toList.flatMap(hex => hexadecimalMapping(hex).toCharArray)
+  ...
+```
+
+This will produce the input that we have seen in the puzzle description:
+
+`110100101111111000101000`
+
+Since we've got that we can start defining our function for decoding packets. We
+first define parsing of elements common to all the packets, which is the version
+and the type ID. Based on the type id we can see how we need to parse the rest
+of the data.
+
+```scala
+type BinaryData = List[Char]
+
+// helper function to read binary data 01101 to decimal 13
+def toInt(chars: BinaryData): Int =
+  Integer.parseInt(chars.mkString, 2)
+
+// helper function to read binary data 01101 to decimal 13, but in a Long format
+def toLong(chars: BinaryData): Long =
+  java.lang.Long.parseLong(chars.mkString, 2)
+
+def readLiteralBody(input: BinaryData): (Long, BinaryData) = ???
+def readOperatorBody(input: BinaryData): (List[Packet], BinaryData) = ???
+
+def decodePacket(packet: BinaryData): (Packet, BinaryData) =
+  val (versionBits, rest) = packet.splitAt(3)
+  val version = toInt(versionBits)
+  val (typeBits, body) = rest.splitAt(3)
+  val tpe = toInt(typeBits)
+
+  tpe match
+    case 4 =>
+      val (value, remaining) = readLiteralBody(body, Nil)
+      (Packet.Literal(version, value), remaining)
+    case otherTpe =>
+      val (values, remaining) = readOperatorBody(body)
+      (Packet.Operator(version, otherTpe, values), remaining)
+  end match
+end decodePacket
+```
+
+We use the function `splitAt`, which gives us the ability to split the input
+into the part that we need, for example 3 bits for version, and the rest of the
+packet data. This way we can read the version and typeId, pattern match on the
+latter and use proper logic for reading in each case. We can then create our new
+structures. We also defined a helper type `BinaryData`, since we will be using
+it throughout the puzzle. What remains is defining `readLiteralBody` and
+`readOperatorBody`. You might notice that we return additional `BinaryData` from
+each function. This is because we will be later able to use it to analyze the
```

Comment:
You sometimes use “analyze” and sometimes “analyse”. I wonder if we should settle on one style.

Reply:
You should definitely settle on one style. As developers we often settle on the US spelling, so "analyze".

Reply:
Settled on `analyze` :sweat_smile: